### DISCUSSION TITLE
Error when var file includes undeclared variables

### DISCUSSION THREAD

**Discussion by rpdelaney** [terraform-linters/tflint#2080] [Ideas]
## Introduction

Terraform gives appropriate warnings for values found in tfvars without a corresponding variable in the configuration:

![2024-06-28_09-19-19](https://github.com/terraform-linters/tflint/assets/1139517/4510b3d6-50ec-488a-a7eb-be7a7fbd67e6)

## Proposal

tflint should also raise errors for the same problem.

## References

N/A

**Comment by bendrucker** [comment]
Duplicate of #1877
Resolved #1941

```sh
tflint --var 'foo=bar'
```

```
Failed to parse variables; <value for var.foo>:1,1-1: Value for undeclared variable; A variable named "foo" was assigned, but the root module does not declare a variable of that name.:

Error: Value for undeclared variable

  on <value for var.foo> line 1:
  (source code not available)

A variable named "foo" was assigned, but the root module does not declare a variable of that name.
```

**Comment by rpdelaney** [comment]
@bendrucker I'm sorry, but I'm not sure I understand how this is a solution for me. As a user I want to use `tflint` in CI to scan for tfvars settings that do not have a corresponding variable. Adding extra command-line switches does not help because that would require me to do the analysis myself, wouldn't it?

**Comment by rpdelaney** [comment]
Reading the tickets you mentioned, they seem to describe a very different issue than the new feature I'm proposing.

**Comment by bendrucker** [comment]
Hmm, I looked at `--var` as a shortcut but it turns out `--var-file` behaves differently and does not error. 

```sh
echo 'foo = "bar"' > terraform.tfvars
tflint --var-file terraform.tfvars
# nothing...
```

#1941 says:

> Note that the environment variables TF_VAR_* are not subject to this check. This is the same behavior as Terraform.

Given that the goal is to error in the cases where Terraform prints a warning, this is looking more like a bug/inconsistency.

**Comment by wata727** [comment]
See https://github.com/hashicorp/terraform/blob/v1.9.0/internal/backend/backendrun/unparsed_value.go#L58-L61

Reading the above comment, it seems like Terraform intentionally makes this a warning rather than an error. There is room for discussion on whether TFLint should actively raise an error or conservatively keep the same behavior as Terraform.

Personally, I think it might be better to provide it as a rule that can be enabled when needed.

**Comment by rpdelaney** [comment]
If `tflint` returns a message with the same log level for this as e.g. when a local variable is declared but unused, that would feel internally consistent, no?

**Comment by wata727** [comment]
It's consistent, but I'm not sure if it helps. Unlike Terraform warnings, TFLint logs are intended for debugging, so they may not be useful for finding typos or other problems.

In your use case, does this help?

**Comment by bendrucker** [comment]
Hmm ok, definitely not a bug on a third pass...

```sh
terraform plan -var foo=bar
```

```
│ Error: Value for undeclared variable
│
│ A variable named "foo" was assigned on the command line, but the root module
│ does not declare a variable of that name. To use this value, add a "variable"
│ block to the configuration.
```

```sh
terraform plan -var-file <(echo foo=bar)
```

```
No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration and
found no differences, so no changes are needed.
```

TFLint is consistent with Terraform.

As far as a rule, the struggle will be referencing the tfvars files. Because they are not part of the module, you'd have to enumerate every tfvars file in the rule config. And then your config is only valid a specific module.

**Reply by rpdelaney** [reply]
This test case:
```console
terraform plan -var foo=bar
```
does not duplicate my problem. The (potentially missing) variables are being referenced in the terraform code on disk, not via command-line.

**Comment by velkovb** [comment]
Just upgraded my Terragrunt version which includes a newer TFLint version and this is a breaking behaviour. We inject a lot of automatic variables that not all modules have, especially third party ones. This really should be a rule that can be toggled.

**Reply by bendrucker** [reply]
How does "injection" work and not cause Terraform to fail in the same way? And why doesn't whatever approach you use for Terraform work for TFLint (e.g., using environment variables or var files)?

**Reply by velkovb** [reply]
Terraform displays a warning, not a failure when you pass values for undeclared variables to it. I would say this is specific to Terragrunt usage. It passes each defined input in the `TF_VAR_variablename` format with value as json. We use a monorepository and have a lot of autogenerated metadata that is useful in our home-grown projects but not all of them use it and we sometimes call public modules directly, which don't have such variables defined. 

So far we haven't had any issues until this upgrade.

**Reply by bendrucker** [reply]
https://github.com/terraform-linters/tflint/discussions/2080#discussioncomment-9920370

`terraform plan` errors when you pass `-var` with a non-existent variables.

Why not also use `TF_VAR_*` with TFLint for your variables that may not exist?

**Reply by velkovb** [reply]
I am a bit confused with `var-file` and `TF_VAR_` I can provide values for already declared variables. My issues is that they are not declared and I cannot declare them dynamically.

I rarely use Terraform directly so hadn't really tested it but I see that it is a warning only if I pass a `.tfvars` files with values for undeclared variables.

But Terragrunt uses the `TF_VAR_*` approach, not sure why now it is throwing those errors.

**Reply by velkovb** [reply]
I think I get it now. Terragrunt does the following:
`tflint --config ./.terragrunt-cache/ellCj9USNaAqucT2-U5SmL6rf1A/Gu7P37gZ1L3NBihh7GJE9nHhW38/.tflint.hcl --chdir ./.terragrunt-cache/ellCj9USNaAqucT2-U5SmL6rf1A/Gu7P37gZ1L3NBihh7GJE9nHhW38 --var=test=foo` 

and before the upgrade it was on TFLint 0.47.0 and this functionality was introduced in 0.50.0. Will open an issue on Terragrunt repo to see what can be done about it. Thank you for you time :)

**Reply by bendrucker** [reply]
Sure! It sounds like moving to passing variables as `TF_VAR` environment variables from Terragrunt would work. Depending on when the integration was added, it's possible that feature didn't exist yet.

The idea behind matching Terraform behavior is that in the ideal case things "just work" when users call TFLint ahead of a `terraform plan` in CI. A config option would get the job done but users would have to know to look for it.

**Reply by velkovb** [reply]
Yes, now I understand it is not a TFLint issue but rather a Terragrunt one :) https://github.com/gruntwork-io/terragrunt/issues/3669

**Reply by bendrucker** [reply]
Great, feel free to mention me in that issue or a corresponding PR if you need input
